crates/cadenza-eval/src/typeinfer.rs

```diff
@@ -0,0 +1,921 @@
+//! - Records with structural typing
+//! - Union types
+//!
+//! The type inference algorithm operates in three phases:
+//! 1. Constraint Generation: Walk the AST and generate type equations
+//! 2. Constraint Solving: Use unification to solve the equations
+//! 3. Generalization: Introduce quantifiers at let-bindings for polymorphism
+
+use crate::{
+    diagnostic::{Diagnostic, DiagnosticKind, Result},
+    interner::InternedString,
+    value::Type,
+};
+use cadenza_syntax::span::Span;
+use rustc_hash::FxHashMap;
+use std::fmt;
+
+/// A type variable used during type inference.
+///
+/// Type variables are placeholders that get unified with concrete types
+/// during the inference process.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct TypeVar(u32);
+
+impl TypeVar {
+    /// Creates a new type variable with the given ID.
+    pub fn new(id: u32) -> Self {
+        Self(id)
+    }
+
+    /// Gets the ID of this type variable.
+    pub fn id(&self) -> u32 {
+        self.0
+    }
+}
+
+impl fmt::Display for TypeVar {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // Use lowercase Greek letters for display
+        let letters = [
+            'α', 'β', 'γ', 'δ', 'ε', 'ζ', 'η', 'θ', 'ι', 'κ', 'λ', 'μ', 'ν', 'ξ', 'ο', 'π', 'ρ',
+            'σ', 'τ', 'υ', 'φ', 'χ', 'ψ', 'ω',
+        ];
+        if self.0 < letters.len() as u32 {
+            write!(f, "{}", letters[self.0 as usize])
+        } else {
+            write!(f, "t{}", self.0)
+        }
+    }
+}
+
+/// A type that may contain type variables during inference.
+///
+/// This is separate from the runtime `Type` enum to avoid polluting it with
+/// inference-specific details.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum InferType {
+    /// A concrete type (no type variables).
+    Concrete(Type),
+    /// A type variable to be inferred.
+    Var(TypeVar),
+    /// A function type with argument types and return type.
+    Fn(Vec<InferType>, Box<InferType>),
+    /// A list type with element type.
+    List(Box<InferType>),
+    /// A record type with field names and types.
+    Record(Vec<(InternedString, InferType)>),
+    /// A tuple type with element types.
+    Tuple(Vec<InferType>),
+    /// A union type (one of several alternatives).
+    Union(Vec<InferType>),
+    /// A polymorphic type with quantified variables.
+    Forall(Vec<TypeVar>, Box<InferType>),
+    /// A quantity type with a unit dimension.
+    Quantity {
+        /// The numeric value type (Integer or Float).
+        value_type: Box<InferType>,
+        /// The dimension (for now we'll store as a placeholder).
+        /// TODO: Integrate with unit system
+        dimension: InternedString,
+    },
+}
+
+impl InferType {
+    /// Converts this inferred type to a concrete runtime type.
+    ///
+    /// Returns an error if the type still contains unresolved type variables.
+    pub fn to_concrete(&self) -> Result<Type> {
```

Comment:
It would be nice if we could cache a lot of this, especially since it's recursive